x-pack/plugins/features/common/sub_feature.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { RecursiveReadonly } from '@kbn/utility-types';
+import { FeatureKibanaPrivileges } from './feature_kibana_privileges';
+
+/**
+ * Configuration for a sub-feature.
+ */
+export interface SubFeatureConfig {
+  /** Display name for this sub-feature */
+  name: string;
+
+  /** Collection of privilege groups */
+  privilegeGroups: SubFeaturePrivilegeGroupConfig[];
+}
+/**
+ * Configuration for a sub-feature privilege group.
+ */
+export interface SubFeaturePrivilegeGroupConfig {
+  /**
+   * The type of privilege group.
+   * - `mutually_exclusive`::
+   *     Users will be able to select at most one privilege within this group.
+   *     Privileges must be specified in descending order of permissiveness (e.g. `All`, `Read`, not `Read`, `All)
+   * - `independent`::
+   *     Users will be able to select any combination of privileges within this group.
+   */
+  groupType: 'mutually_exclusive' | 'independent';
+
+  /**
+   * The privileges which belong to this group.
+   */
+  privileges: SubFeaturePrivilegeConfig[];
+}
+
+/**
+ * Configuration for a sub-feature privilege.
+ */
+export interface SubFeaturePrivilegeConfig
+  extends Omit<FeatureKibanaPrivileges, 'excludeFromBasePrivileges'> {
+  /**
+   * Identifier for this privilege. Must be unique across all other privileges within a feature.
+   */
+  id: string;
+
+  /**
+   * The display name for this privilege.
+   */
+  name: string;
+
+  /**
+   * Denotes which Primary Feature Privilege this sub-feature privilege should be included in.
+   * `read` is also included in `all` automatically.
+   */
+  includeIn: 'all' | 'read' | 'none';
+}
+
+export class SubFeature {
+  constructor(protected readonly config: RecursiveReadonly<SubFeatureConfig>) {}
+
```

Comment:
Valid question! On its own, it's honestly not all that useful, but I took this approach so that consumers had a consistent way to access `Feature`s and `SubFeature`s. The security plugin extends both of these models in `x-pack/plugins/security/public/management/roles/model` to add security-specific constructs